### ISSUE TITLE
Null reference at NHibernate.Util.AsyncReaderWriterLock.ReadLock()

### ISSUE THREAD

**Issue by michalstepien** [nhibernate/nhibernate-core#2468]
Hi, 
after upgrade NHibernate 5.2.7 -> 5.3.1, I have this error:

```
[NullReferenceException]
Object reference not set to an instance of an object.
   at NHibernate.Util.AsyncReaderWriterLock.ReadLock()
   at NHibernate.Cache.ReadWriteCache.Get(CacheKey key, Int64 txTimestamp)
   at NHibernate.Event.Default.DefaultLoadEventListener.LoadFromSecondLevelCache(LoadEvent event, IEntityPersister persister, LoadType options)
   at NHibernate.Event.Default.DefaultLoadEventListener.DoLoad(LoadEvent event, IEntityPersister persister, EntityKey keyToLoad, LoadType options)
   at NHibernate.Event.Default.DefaultLoadEventListener.Load(LoadEvent event, IEntityPersister persister, EntityKey keyToLoad, LoadType options)
   at NHibernate.Event.Default.DefaultLoadEventListener.ProxyOrLoad(LoadEvent event, IEntityPersister persister, EntityKey keyToLoad, LoadType options)
   at NHibernate.Event.Default.DefaultLoadEventListener.OnLoad(LoadEvent event, LoadType loadType)
   at NHibernate.Impl.SessionImpl.FireLoad(LoadEvent event, LoadType loadType)
   at NHibernate.Impl.SessionImpl.InternalLoad(String entityName, Object id, Boolean eager, Boolean isNullable)
   at NHibernate.Type.EntityType.ResolveIdentifier(Object id, ISessionImplementor session)
   at NHibernate.Collection.Generic.PersistentGenericSet`1.InitializeFromCache(ICollectionPersister persister, Object disassembled, Object owner)
   at NHibernate.Cache.Entry.CollectionCacheEntry.Assemble(IPersistentCollection collection, ICollectionPersister persister, Object owner)
   at NHibernate.Event.Default.DefaultInitializeCollectionEventListener.Assemble(CacheKey ck, Object ce, ICollectionPersister persister, ISessionImplementor source, IPersistentCollection collection, Object collectionKey, Boolean alterStatistics)
   at NHibernate.Event.Default.DefaultInitializeCollectionEventListener.InitializeCollectionFromCache(Object collectionKey, ICollectionPersister persister, IPersistentCollection collection, ISessionImplementor source)
   at NHibernate.Event.Default.DefaultInitializeCollectionEventListener.OnInitializeCollection(InitializeCollectionEvent event)
   at NHibernate.Impl.SessionImpl.InitializeCollection(IPersistentCollection collection, Boolean writing)
   at NHibernate.Collection.AbstractPersistentCollection.Initialize(Boolean writing)
   at NHibernate.Collection.AbstractPersistentCollection.ReadElementExistence[T](T element, Nullable`1& existsInDb)
   at NHibernate.Collection.Generic.PersistentGenericSet`1.Contains(T item)
```

The error occurs during high load applications.

**Comment by fredericDelaporte** [comment]
The offending line of code can only be:
https://github.com/nhibernate/nhibernate-core/blob/c3ac6a0fc599d2a384407ac44a9b5eff89d2149e/src/NHibernate/Util/AsyncReaderWriterLock.cs#L90
I think it should just be switched to:
```c#
_waitingReadLockSemaphore?.Wait();
```
If the waiting lock has already been switched out, it means the read can go on.

**Comment by maca88** [comment]
> The offending line of code can only be:
_waitingReadLockSemaphore.Wait(); 

Correct, what happend in this scenario:
```C#
public Releaser ReadLock()
{
	if (CanEnterReadLock())
	{
		return _readerReleaser;
	}

	// Before the below line is executed ExitWriteLock is called, which set _waitingDisposalSemaphore to null
	// and increases the _currentReaders by at least one as the above CanEnterReadLock returned false. If WriteLock
	// is called again it will wait due to _currentReaders is greater than zero. So there are two solutions, either
	// skip calling _waitingReadLockSemaphore when is null or add an output parameter for CanEnterReadLock in order
	// to obtain _waitingReadLockSemaphore
	_waitingReadLockSemaphore.Wait();

	return _readerReleaser;
}
```
to me adding an output parameter is more correct, which I did in #2469.

**Comment by bahusoid** [comment]
> to me adding an output parameter is more correct

Just a question: are you sure when Wait method is called it can't be already disposed?

**Comment by maca88** [comment]
> are you sure when Wait method is called it can't be already disposed?

The semaphore is disposed (`DisposeWaitingSemaphore`) in `WriteLock` only when `_currentReaders` is zero, which in this case won't be:
https://github.com/nhibernate/nhibernate-core/blob/c3ac6a0fc599d2a384407ac44a9b5eff89d2149e/src/NHibernate/Util/AsyncReaderWriterLock.cs#L43-L59

In this case `CanEnterWriteLock` will return `true`, but `waitForReadLocks` will also be `true` due to `_currentReaders` being greater than zero, so the disposal will wait until the read lock is released.
So yes, I am pretty sure it can't be disposed when called.

**Comment by hazzik** [comment]
Fixed in #2469

**Comment by Gonnagle** [comment]
I run into bunch of integration tests still failing on following error with 5.3.3
```
System.ObjectDisposedException : Cannot access a disposed object.
Object name: 'AsyncReaderWriterLock'.
   at NHibernate.Util.AsyncReaderWriterLock.AssertNotDisposed()
   at NHibernate.Util.AsyncReaderWriterLock.CanEnterWriteLock(Boolean& waitForReadLocks)
   at NHibernate.Util.AsyncReaderWriterLock.WriteLock()
   at NHibernate.Cache.ReadWriteCache.Lock(CacheKey key, Object version)
   at NHibernate.Action.CollectionAction.BeforeExecutions()
   at NHibernate.Engine.ActionQueue.PrepareActions[T](List`1 queue)
   at NHibernate.Event.Default.AbstractFlushingEventListener.PerformExecutions(IEventSource session)
   at NHibernate.Event.Default.DefaultFlushEventListener.OnFlush(FlushEvent event)
   at NHibernate.Impl.SessionImpl.Flush()
   at NHibernate.Impl.SessionImpl.BeforeTransactionCompletion(ITransaction tx)
   at NHibernate.Transaction.AdoTransaction.Commit()
```

Reverting to 5.2.7 seemed to fix the issue so will go with that as a workaround for now. Could definitely be that there is just something fishy going on our code causing this, but unfortunately can't dive much deeper right now at least.
